### PR TITLE
Turn exclude_border off in peak_local_max by default

### LINKED ISSUE #2407: Better documentation for `peak_local_max`s `exclude_border` boolean variant
## Description

The documentation of [`peak_local_max`](http://scikit-image.org/docs/dev/api/skimage.feature.html#skimage.feature.peak_local_max) says nothing about the boolean variant of the `exclude_border` parameter, although it defaults to `True`. So it's unclear from the documentation what that is supposed to mean.

By looking at the corresponding [code](https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/peak.py#L115), one can see that `min_distance` is used for `exclude_border` when it is passed `True`.

**First improvement:** Update the corresponding documentation to make this invariant clear.

Additionally, I think the default value of `True` might lead to subtle problems, that people might not be aware of. For instance, I wasn't aware of this fact until now and just found it by visual inspection. My mental model of NMS doesn't seem to fit the default implementation of `peak_local_max`, which I find problematic.

**Second improvement:** Don't use `True` as default for `exclude_border`, instead use `0`.

### PR DESCRIPTION
**Pull request by b52**
Fixes #2407

### CODE REVIEW

File: skimage/feature/peak.py [hunk 1]
```diff
@@ -48,9 +48,10 @@ def peak_local_max(image, min_distance=1, threshold_abs=None,
         the minimum intensity of the image.
     threshold_rel : float, optional
         Minimum intensity of peaks, calculated as `max(image) * threshold_rel`.
-    exclude_border : int, optional
+    exclude_border : int | bool, optional
```
Comment (soupault) [inline skimage/feature/peak.py]: `int | bool` -> `int or bool` (see https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)

Comment (b52) [inline skimage/feature/peak.py]: @soupault Done

**Comment by sciunto** [comment]
Shouldn't we follow a deprecation path?

**Comment by b52** [comment]
I don't think so. First, there is nothing going to be removed nor added in place of it. Second, the change of the default value only might cause more peaks popping up for unaware users, which is most certainly better than the other way around.

So I think an appropriate remark in the change log is sufficient.

**Comment by sciunto** [comment]
I'm hesitant. At least, I would not include this patch in 0.13 to have enough time to get feedbacks from users who test the dev version.

**Comment by jni** [comment]
Uuuuh, other @scikit-image/core devs can chime in, but I'm pretty sure this requires a deprecation path. This is required whenever the output of a function will change for the exact same input. When people update their environment, they should not be left to wonder which of the updates changed their results (NumPy, SciPy, pandas, scikit-image could all be culprits in a complex pipeline).

**Comment by b52** [comment]
@jni It's not the exact same input, since the default value changed and thus a different input.

Nonetheless, how should one proceed in case a deprecation path is required?

**Comment by JDWarner** [comment]
For the record, when I first combined two separate peak-finding functions into what is now `peak_local_max` back in 2013 or so, I thought this setting was the _worst_.  It's very unintuitive, and IMO should absolutely be disabled by default.  At the time, I didn't have the confidence to say so; the test suite assumed this behavior and there was no deprecation path.

What's changed in 4 years?  Nothing, except we've led a lot of users astray who had peak(s) near the edge of their images and didn't change the default.  Worse, our default links the blind edge to `min_distance` which is extremely unintuitive and problematic.

At issue, really, is this:

If a function saying "hey I find peaks!" _which actually ignores any peaks in a large subset of the image_ with default settings hostile to users?  I'd argue yes.  Further, I'd argue that anyone who would be tripped up by changing this default _in the vast majority of cases was assuming incorrect behavior_.  Users who did change the default appropriately would be unaffected.

Coming at this from another angle - can anyone think of a single common use case where the border exclusion would actually be desired?  I can make up very niche reasons this would be nice, but none of them are good reasons.  And all of them could be handled by doing peak finding on a sliced down input, e.g., passing `im[4:-4, 4:-4]`.

In summary, I think this is user hostile behavior and a bug.  It should be fixed.  There's no easy deprecation pathway, so a documented breaking change is warranted.

**Comment by b52** [comment]
That makes me wonder why it was included in the first place @JDWarner? If there were no legit reasons or use cases, it seems kinda odd to add a feature just for the feature.

**Comment by JDWarner** [comment]
@b52 Can't say; I'm unsure of the history before that point, as harmonizing the peak finding was one of my earliest contributions.  Maybe there's a decent reason for it I'm missing entirely.

**Comment by stefanv** [comment]
How about: removing the setting and raising a warning for two releases, unless it is specified and set to False.

**Comment by stefanv** [comment]
@b52 Are you still interested in working on this issue?  @JDWarner @jni Are you both happy with the removal of this option, as suggested above?

**Comment by jni** [comment]
@stefanv et al

Actually I suggest closing this PR. Some points:

- Sorry that I missed the discussion about *why* have that option at all, but `exclude_border` absolutely makes sense in many contexts. Here's a simple example: 

screen shot 2018-05-01 at 8 15 23 pm

The point I'm trying to illustrate is that one cannot be sure whether a maximum on the border is actually a local maximum or rather a truncated upward slope. In fact, the second option is much more probable if you are sampling peaks that are wider than three pixels. So if you don't exclude the borders, you will mostly get junk on the borders, in proportion to how wide your peaks are. And @JDWarner, truncating the image does not actually alleviate this problem because then you will get junk on the inner border — and real peaks in the intervening space will be mis-localized!

In other examples, such as object counting that correctly accounts for edge effects, you do want peaks on the borders, but only half the borders.

- Apart from this, we have been [having many issues](https://github.com/scikit-image/scikit-image/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+peak_local_max) with `peak_local_max` and are currently exploring alternate algorithms, see #3022, that will probably render work specifically on `peak_local_max` obsolete.

- Apart from that, as I mentioned in the original comments, changing a default value *should* follow a deprecation path, because people running `peak_local_max(image)` anywhere in their tests will now have completely different output. The case of default value changes is specifically addressed in our [contributing guidelines](http://scikit-image.org/docs/dev/contribute.html#deprecation-cycle). 

To summarize my view:

a) the option should be maintained going forward
b) we can argue about the default value (this should take place in #3022 or #3016 probably), but I would argue that the default should be to exclude a 1 pixel border.
c) any change in behavior should follow the standard deprecation path. `peak_local_max` may have problems but based on help requests I predict it's among our most used functions.

@b52 thanks for your PR, and I apologize if I sound rude above... I just want to make my case clear and also not send you down the rabbit hole of deprecation procedures when the function itself might be deprecated soon.

**Comment by stefanv** [comment]
I think @jni's point is that those points on the border that seem like maxima may very well not be, had we been able to sample "further afield".  On the other hand, if you define a maximum as a pixel with no greater neighbor, then these count.  So, yes, this should remain an option, and it is not clear what to make the default—either option seems fine, as long as it can easily be disabled.

The current status is just weird, since it defaults to "True", which doesn't seem right—should be 1 or 0.  The docstring states:

```
exclude_border : int, optional
    If nonzero, `exclude_border` excludes peaks from
    within `exclude_border`-pixels of the border of the image.
```

So, what happens to `True` here?  Does it become 1?  If so, then your suggestion is already implemented, @jni, but should be clarified in the docstring.

**Comment by jni** [comment]
@stefanv if `True`, then the value evaluates to `min_distance`! I thought we had merged a PR that made this clear, but clearly this is not the case.
